ini/parser.go

```diff
@@ -0,0 +1,317 @@
+func (p *Parser) SaveToFile(path string) error {
+	// saves the parsed dict as ini file in the specified path after converting to string
+	file, err := os.Create(path)
+	defer file.Close()
+
+	if err == nil {
+		_, err = file.WriteString(p.ToString())
+	}
+
+	return err
+}
+
+func (p *Parser) FromString(content string) error {
+	// parse the ini content as :
+	// [section1: map[key1: value1 key2: value2 ...] section2: map[key1: value1 key2: value2 ...] ...]
+
+	var err error
+
+	p.parsedMap = INIMap{}
+	var key string
+	var val string
+	var parent string
+
+	// for parents
+	var newParent bool = false
+
+	// read content lines
+	scanner := bufio.NewScanner(s.NewReader(content))
+	for scanner.Scan() {
+		s.ReplaceAll(scanner.Text(), "\n", "")
+		s.ReplaceAll(scanner.Text(), "\r", "")
+
+		line := scanner.Text()
+
+		if len(line) > 0 {
+			// parse sections
+			if string(line[0]) == "[" && string(line[len(line)-1]) == "]" {
+				// check number of opened and closed sections []
+				if s.Count(line, "[") == 1 && s.Count(line, "]") == 1 {
+					parent = line[1 : len(line)-1]
+					p.addParent(parent)
+					newParent = true
+				} else {
+					err = InvalidSection
+				}
+
+				// parse sections values
+			} else if newParent && s.Count(line, "=") == 1 && (!contains([]string{"", "=", " "}, string(line[0])) && !contains([]string{"", "=", " "}, string(line[len(line)-1]))) {
+				if s.Contains(line, " = ") {
+					splitted := s.Split(line, " = ")
+					key = splitted[0]
+					val = splitted[1]
+				} else if s.Contains(line, "=") {
+					splitted := s.Split(line, "=")
+					key = splitted[0]
+					val = splitted[1]
+				}
+
+				p.add(parent, key, val)
+
+				// parse comment lines
+			} else if string(line[0]) == ";" {
+				continue
+
+				// invlid content
+			} else {
+				err = InvalidContent
+			}
+
+			// parse empty
+		} else if s.Trim(line, " ") == "" {
+			continue
+
+			// invlid content
+		} else {
+			err = InvalidContent
+		}
+	}
+
+	return err
+}
+
+func (p *Parser) ToDict() INIMap {
+	// Get the parsed dictionary as :
+	// [section1: map[key1: value1 key2: value2 ...] section2: map[key1: value1 key2: value2 ...] ...]
+
+	return p.parsedMap
+}
+
+func (p *Parser) GetSections() ([]string, error) {
+	// returns a list of all sections --> [section1 section2 ...]
+
+	var err error
+	var sections []string
+
+	for parent := range p.parsedMap {
+		sections = append(sections, parent)
+	}
+
+	// if no data in the parsed dict
+	if len(sections) == 0 {
+		err = NoParsedData
+	}
+
+	return sections, err
+}
+
+func (p *Parser) GetSection(sectionKey string) (Entry, error) {
+	// returns a dictionary for the section given --> map[key1: val1 key2: val2 ....]
+
+	sections, err := p.GetSections()
+	var section Entry
+
+	if err == nil && !contains(sections, sectionKey) {
+		err = SectionNotExist
+	} else {
+		section = p.parsedMap[sectionKey]
+	}
+	return section, err
+}
+
+func (p *Parser) GetOptions(sectionKey string) ([]string, error) {
```

Comment:
I'd make it okay to return an empty list without an error, but if we will force the sections to have values, it's fine. let's also add the doc